Add Adaptive Support

garrettwrong opened the pull request:
Closes #431 - Some notes in issue.

One point of discussion is location.  I had considered making this a method of `Image` class.  Also it is located under `denoising` based on the legacy location, but perhaps there is a better place.

garrettwrong committed:
Add adaptive support method and unit tests

garrettwrong committed:
Cleanup adaptive support code for review

janden commented on src/aspire/denoising/adaptive_support.py:
```diff
@@ -0,0 +1,107 @@
+    Images are scaled in real space so that the noise variance in both
+    real and Fourier domains is similar.
+
+    Returns c_limit (support in Fourier space), R_limit (support in real space).
```
Should make clear that these are radii.

garrettwrong replied:
adjusted comment

janden commented on src/aspire/denoising/adaptive_support.py:
```diff
@@ -0,0 +1,107 @@
+logger = logging.getLogger(__name__)
+
+
+def adaptive_support(images, energy_threshold=0.99):
```
In the future, it might make more sense for this to take an `ImageSource` as input, since we can't always rely on all images being in memory.

garrettwrong replied:
I've naively changed to using an `ImageSource` or subclass.

I have not changed the code to be smart (batch-wise calculations) with the image stack internally.  I think this is a fair compromise to get the interface using the class without investing a lot into code that may not get used near term.

janden replied:
Makes sense.

janden commented on src/aspire/denoising/adaptive_support.py:
```diff
@@ -0,0 +1,107 @@
+
+    # Note that the noise theoretically same, but in practice differ
+    # Take smaller so we do not yield negative variance or PSD later
+    noise_var = min(np.var(img_corner), np.var(imgf_corner))
```
These lines (minus the `np.var(imgf_corner)` part) essentially corresponds to applying a `WhiteNoiseEstimator`, so this should probably be refactored in the future. Right now, that only takes an `ImageSource` as input, but if this is eventually rewritten to take `ImageSource`s as input, it should be fairly straightforward.

garrettwrong replied:
Yep, I replaced with `WhiteNoiseEstimator` after changing to an ImageSource.

janden commented on src/aspire/denoising/adaptive_support.py:
```diff
@@ -0,0 +1,107 @@
+        radial_pspec[i] = np.mean(pspec[(r >= i) & (r < i + 1)])
+
+    # Subtract the noise variance
+    radial_pspec -= noise_var
```
Should probably clip this at zero.

garrettwrong replied:
added clip call

janden commented on src/aspire/denoising/adaptive_support.py:
```diff
@@ -0,0 +1,107 @@
+
+    # Compute the Radial Variance
+    radial_var = np.zeros(N)
+    for i in range(N):
```
Merge this loop with the next to make it clear that they're essentially doing the same thing.

garrettwrong replied:
refactored along with the variance_map change below.

janden commented on src/aspire/denoising/adaptive_support.py:
```diff
@@ -0,0 +1,107 @@
+    # Take smaller so we do not yield negative variance or PSD later
+    noise_var = min(np.var(img_corner), np.var(imgf_corner))
+
+    variance_map = np.var(
```
Since images have already been demeaned, this can be calculated in the same way as the power spectrum below (`np.mean(np.abs(img) ** 2, axis=0)`).

garrettwrong replied:
this returns very different results for gaussian disc, but they look better ... interesting...

janden replied:
Maybe I misunderstood what this is doing then. Could it be the division by `N - 1` instead of `N`?

garrettwrong replied:
In the original code they were using an odd syntax of the var call.  It was afaict not documented, but corresponded to one less dof.  I'm thinking it was not correct anyway. Assuming there is time we can talk about it tomorrow.

janden commented on src/aspire/denoising/adaptive_support.py:
```diff
@@ -0,0 +1,107 @@
+    R = np.arange(N, dtype=int)
+
+    # Calculate cumulative energy
+    cum_pspec = np.zeros(N)
```
`np.cumsum(radial_pspec * c)`?

garrettwrong replied:
yes, that's cleaner, done.

janden commented on src/aspire/denoising/adaptive_support.py:
```diff
@@ -0,0 +1,107 @@
+    imgf = fft.centered_fft2(images.asnumpy())
+
+    # demean
+    img = images.asnumpy() * L
```
This is odd.

garrettwrong replied:
I thought so too, but at the time I assumed the original code was more mature than it turned out.  I'll try to back out why that might be there.

janden replied:
The only thing I can think of is rescaling the variance estimate (effectively multiplying it by `L ** 2`), but it's not clear to me why this is needed.

garrettwrong replied:
It's not needed.  I think it was an artifact in the old code relating to expectations of the input data.  The old code if it was ever run was getting a bespoke array from class averaging code. I've removed.

janden commented on src/aspire/denoising/adaptive_support.py:
```diff
@@ -0,0 +1,107 @@
+    # Note legacy code *=L for Fourier limit,
+    #   but then only uses divided by L... so just removed here.
+    # This makes it consistent with Nyquist, ie [0, .5]
+    c_limit = c[np.argwhere(cum_pspec > energy_threshold)[0, 0]]
```
`np.argmax(cum_pspec > energy_threshold)`? Don't know if that makes it clearer, but this will only get the index of the first true element.

garrettwrong replied:
much clearer yes, haven't used that one much. Will try to remember it.

janden commented on tests/test_adaptive_support.py:
```diff
@@ -0,0 +1,54 @@
+            np.load(os.path.join(DATA_DIR, "sim_images_with_noise.npy"))
+        )
+
+    def test_adaptivate_support(self):
```
Maybe add synthetic tests that test images of centered Gaussians? These have the advantage that we can calculate the correct `R` and `c` analytically (using `scipy.special.erf`).

garrettwrong replied:
That's a great idea, thanks. I'll do this then hit the other items.

garrettwrong replied:
I was able to make some progress here.  The real support seems to be working okay.   For a higher res image I am able to roughly estimate 1,2,3 standard deviations.  Example for centered gaussian disc with stddev 128 the code might return ~240 pixels for threshold 0.966 (two std devs).

I think may be an issue for the Fourier support values, or I'm deriving it wrong.  (By it, I mean the corresponding fourier transformed gaussian). TBD.

garrettwrong replied:
I had the correct gaussian, talked about yesterday.  I've also tried two other functions from our meeting yesterday (`circ` and `hankel`).  I was able to incorporate the circ with the gaussian for one of the tests.

I was hoping since Hankel should be it's own transform that I could use this for a crude equivalence between Real and Fourier space, and test the F support, but it doesn't seem right to me.  Trying to figure that out now.

garrettwrong replied:
Concretely, when I fftshift(fft2(hankel(...))), the resulting image is not as close to original as I thought it would be.

garrettwrong replied:
I added two more functions for generating some test data to `misc`.  One for `circ` and another for `hankel`.  Now there are three synthetic tests:

1) Test that Fourier support of a Gaussian is similar to normal distribution.
2) Test that real support of a `hankel` image (1/(x**2 + y**2) and its Fourier transform are similar.
3) Test that Gaussian plus various `circ` radius achieve reasonable support values for 1,2,3,inf std deviations. Pretty loose, but looks that we are closer to the right std dev threshold.

janden replied:
I added some comments in the new review. Resolving this one.

garrettwrong committed:
adjust comment - radius

garrettwrong committed:
better use of numpy, improve comments

garrettwrong committed:
fix typo/bug std/var

garrettwrong committed:
stashing updates, need to find simple way to test F support

garrettwrong committed:
use ImageSource and merge two unit tests

garrettwrong committed:
incorporate testing updates/suggestions

janden commented on src/aspire/utils/misc.py:
```diff
@@ -109,3 +111,102 @@ def sha256sum(filename):
+    Note for odd resolutions center is the grid point 0,
+    while for even resolutions center is first pixel (grid point 0.5).
+
+    :param resolution: The height and width of returned array (pixels)
```
Should stick with `size` or similar terminology for these things. Resolution often means other things which can be confusing.

garrettwrong replied:
changed.

janden commented on src/aspire/utils/misc.py:
```diff
@@ -109,3 +111,102 @@ def sha256sum(filename):
+    Ll = -(resolution - 1) // 2 + (resolution - 1) % 2
+    # # Upper
+    Lu = resolution // 2 + 1
+    (Y, X) = np.mgrid[Ll:Lu, Ll:Lu]
```
Use `grid_2d`?

garrettwrong replied:
Okay, converted. Slightly different convention re even case.

janden commented on src/aspire/utils/misc.py:
```diff
@@ -109,3 +111,102 @@ def sha256sum(filename):
+
+    Default is a centered disc of spread=peak=1.
+
+    Note for odd resolutions center is the grid point 0,
```
Isn't the center at zero for all sizes?

garrettwrong replied:
is zero a grid point or a pixel... I'll convert to grid_2d per your other comment so we have the same convention.  I didn't think it was terribly important for the ut helper functions, but you are right, they should all be consistent.

janden replied:
:pray:

janden commented on src/aspire/utils/misc.py:
```diff
@@ -109,3 +111,102 @@ def sha256sum(filename):
+    return circ.astype(dtype)
+
+
+def hankel(resolution, x0=0, y0=0, peak=1, dtype=np.float64):
```
Why the name `hankel`? Something like `inverse_r` might be more appropriate. It is also a little concerning since this function has a singularity at zero. Would something like `1 / sqrt(1 + x**2 + y**2)` work?

garrettwrong replied:
trying to be consistent with the documentation from Wiki, which labels this as the 2d case of the Hankel...  I'm not attached to the name... changed.

FWIW I smoothed out the singularity for the grid convention I was using already (before division), but in the grid_2d convention what you propose is probably better. Changed.

janden commented on tests/test_adaptive_support.py:
```diff
@@ -0,0 +1,133 @@
+            # Pass numpy array.
+            _ = adaptive_support(np.empty((10, 32, 32)))
+
+    def test_adaptivate_support_hankel(self):
```
`adaptive`

garrettwrong replied:
typo, my bad.

janden commented on tests/test_adaptive_support.py:
```diff
@@ -0,0 +1,133 @@
+        for d in range(n_disc):
+            discs[d] = discs[d] + circ(resolution, radius=(d + 1) ** 2)
+
+        self.img_src = ArrayImageSource(Image(discs))
```
This is only used in one of the tests. Might be more straightforward to move it to that test.

garrettwrong replied:
used in three tests ,  two test functions... but I see what you mean.. can replace some with dummy data.  Moved into the child functions.

janden replied:
Right I missed the bad input tests. Thanks!

janden commented on tests/test_adaptive_support.py:
```diff
@@ -0,0 +1,133 @@
+
+        for threshold in thresholds:
+            _, r = adaptive_support(img_src, threshold)
+            _, rf = adaptive_support(f_img_src, threshold)
```
I'm confused. Why not get `r` and `rf` from a single call to `adaptive_support` on `img_src`?

garrettwrong replied:
They are not close enough to equivalent after normalization correction (`*size`) to form a unit test.  

I'll try to elaborate below in a related discussion.

janden replied:
Ok closing this then.

janden commented on tests/test_adaptive_support.py:
```diff
@@ -0,0 +1,133 @@
+            c, r = adaptive_support(self.img_src, threshold)
+            # Check closer to this threshold than next
+            logger.info(f"{c} {r} = adaptive_support(..., {threshold})")
+            self.assertTrue(np.abs(r / self.sigma - stddevs) < 0.5)
```
It's not clear to me where the disks come in here?

garrettwrong replied:
Discs are in `self.img_src`.  They are added so that gaussian has some variation.  Otherwise, you are right, in this iteration they are not used in any fundamental way.  

There was a point when I tried using them but they didn't really work well enough for unit testing and I decided to try using the hankel/inverse_r function instead.  Seemed more useful.

The `circ` function is left in mainly for future convenience.

janden replied:
Right, but don't we expect the disks to alter the estimated support in the Fourier domain? They decay much slower than the Gaussian.

garrettwrong replied:
They do, but I had trouble quantifying the Fourier radius result from this method in any meaningful way for a unit test.  Maybe with the change in 03fa86a it's better ... but that is the fourth root of the original formula 😱 ...

janden replied:
Ok then I suggest we get rid of them. It's not obvious how the thresholds of the Gaussian and the disk interact. We could just have a separate test with the disks, but if that's not working, we can just skip it.

garrettwrong replied:
With the reversion in https://github.com/ComputationalCryoEM/ASPIRE-Python/pull/432/commits/b72b3c7c146b0610090b31290aaaf7ef934969d4 , we're not going to get any meaningful output here afaict.

I would like to leave the discs so the code is at least being smoke tested, if that's alright.

garrettwrong committed:
convert to grid_2d

garrettwrong committed:
remove circular import for useless function

garrettwrong committed:
convert resolution to size in code/docs

garrettwrong committed:
minor changes to hankel, rename inverse_r and change how singularity addressed.

garrettwrong committed:
fix typo in adaptivate unit test names

garrettwrong committed:
move test img_src generation into test functions

garrettwrong committed:
tweak formula more for consideration

garrettwrong committed:
revert track normalization term for PSD

garrettwrong committed:
Revert formula to original author's, adds some logging

breaks unit test expectation, marks XFail

janden committed:
Increase accuracy in adaptive_support

janden committed:
Fix adaptive support Gaussian tests

This fixes the three main issues with the Gaussian tests. First, the
formulas for the spatial and frequential cutoffs needed to be adjusted
(there was a missing factor 2π). Second, since we take the square of the
image (and its Fourier transform), we need to correct the sigmas (and
therefore the cutoffs) by dividing by √2. Third, we're not integrating
the 1D Gaussian, but the 2D Gaussian, which is equivalent to integrating
r * exp(-r ** 2). As a result, the thresholds are given by (1 - exp(-r
** 2)), not by the Gaussian error function.

garrettwrong committed:
Merge pull request #445 from janden/432_gaussian_tests

Fix Gaussian tests in #432

garrettwrong committed:
cleanup contrib code

garrettwrong committed:
Remove superflous unit tests
